cassandane/Cassandane/Cyrus/Sieve.pm

```diff
@@ -6158,4 +6158,66 @@ EOF
     }, $res->[0][1]{list}[1]{calendarIds});
 }
 
+sub test_include_cancel_implicit_keep
+    :needs_component_sieve
+{
+    my ($self) = @_;
+
+    xlog $self, "Install a script which includes another";
+    $self->{instance}->install_sieve_script(<<EOF
+require ["include"];
+include "foo";
+EOF
+    );
+
+    xlog $self, "Install a script to be included";
+    $self->{instance}->install_sieve_script(<<EOF
+# This should cancel implicit keep
+discard;
+EOF
+    , name=>'foo');
+
+    xlog $self, "Deliver a message";
+    my $msg = $self->{gen}->generate(subject => "Message 1");
+    $self->{instance}->deliver($msg);
+
+    xlog $self, "Check that no messages are in INBOX";
+    $self->{store}->set_folder('INBOX');
+    $self->check_messages({}, check_guid => 0);
+}
+
+sub test_include_fileinto_implicit_keep_flags
+    :needs_component_sieve
+{
+    my ($self) = @_;
+
+    $self->{store}->set_fetch_attributes(qw(uid flags));
+
+    xlog $self, "Install a script which includes another";
+    $self->{instance}->install_sieve_script(<<EOF
+require ["include"];
+include "foo";
+# Should implicit keep without \Seen, but with \Flagged
```

Comment:
Perl complains about the \S (and the \F) here:

> Unrecognized escape \S passed through at Cassandane/Cyrus/Sieve.pm line 6198.

This is because the here-doc is interpolated (like a `"double quoted string"`).  If I recall correctly, you can force a here-doc to be uninterpolated (like a `'single quoted string'`) by saying `<<'EOF'` instead of `<<EOF`, or something like that.  Or you can just escape the `\`'s the same as you would in any other double-quoted string, like this:

```suggestion
# Should implicit keep without \\Seen, but with \\Flagged
```

Reply:
Fixed